providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -529,19 +563,237 @@ static void ecx_key_adjust(void *key, struct der2key_ctx_st *ctx)
 #  define sm2_d2i_private_key           (d2i_of_void *)d2i_ECPrivateKey
 #  define sm2_d2i_public_key            NULL
 #  define sm2_d2i_key_params            (d2i_of_void *)d2i_ECParameters
+#  define sm2_d2i_PUBKEY                ec_d2i_PUBKEY
+#  define sm2_free                      (free_key_fn *)EC_KEY_free
+#  define sm2_check                     ec_check
+#  define sm2_adjust                    ec_adjust
 
-static void *sm2_d2i_PKCS8(void **key, const unsigned char **der, long der_len,
+static void *sm2_d2i_PKCS8(const unsigned char **der, long der_len,
                            struct der2key_ctx_st *ctx)
 {
     return der2key_decode_p8(der, der_len, ctx,
                              (key_from_pkcs8_t *)ossl_ec_key_from_pkcs8);
 }
-
-#  define sm2_d2i_PUBKEY                (d2i_of_void *)d2i_EC_PUBKEY
-#  define sm2_free                      (free_key_fn *)EC_KEY_free
-#  define sm2_check                     ec_check
-#  define sm2_adjust                    ec_adjust
 # endif
+
+#endif
+
+/* ---------------------------------------------------------------------- */
+
+#ifndef OPENSSL_NO_ML_KEM
+
+typedef enum {
+    NONE_FMT,
+    SEED_FMT,
+    LONG_FMT,
+    ASN1_FMT
+} ML_KEM_PRIV_FMT;
+
+static void *
+ml_kem_d2i_PKCS8(const uint8_t **der, long der_len, struct der2key_ctx_st *ctx)
+{
+    const ML_KEM_VINFO *vinfo;
+    ML_KEM_KEY *key = NULL, *ret = NULL;
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(ctx->provctx);
+    PKCS8_PRIV_KEY_INFO *p8inf = NULL;
+    const unsigned char *p;
+    const X509_ALGOR *alg = NULL;
+    ML_KEM_PRIV_FMT fmt;
+    int plen, ptype, privlen, pairlen;
+    uint16_t keylen;
+
+    /*
+     * The accepted private key format is either the 64-bytes (d, z) seed pair,
+     * or else the optionally DER wrapped octet-string encoding  (one byte tag,
+     * three byte length) of the FIPS 203 expanded |dk| key, optionally
+     * concatenated with the public |ek|.
+     */
+    if ((p8inf = d2i_PKCS8_PRIV_KEY_INFO(NULL, der, der_len)) == NULL)
+        return 0;
+    if (!PKCS8_pkey_get0(NULL, &p, &plen, &alg, p8inf))
+        goto end;
+
+    /* Bail out early if this is some other key type. */
+    if (OBJ_obj2nid(alg->algorithm) != ctx->desc->evp_type)
+        goto end;
+    vinfo = ossl_ml_kem_get_vinfo(ctx->desc->evp_type);
+
+    /* Parameters must be absent. */
+    X509_ALGOR_get0(NULL, &ptype, NULL, alg);
+    if (ptype != V_ASN1_UNDEF) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_UNEXPECTED_KEY_PARAMETERS,
+                       "unexpected parameters with a PKCS#8 %s private key",
+                       vinfo->algorithm_name);
+        goto end;
+    }
+    privlen = (int) vinfo->prvkey_bytes;
+    pairlen = (int) (vinfo->prvkey_bytes + vinfo->pubkey_bytes);
+
+    if (plen == ML_KEM_SEED_BYTES)
+        fmt = SEED_FMT;
+    else if (plen == privlen || plen == pairlen)
+        fmt = LONG_FMT;
+    else if (plen == 4 + privlen || plen == 4 + pairlen)
```

Comment:
4 is "obvious" for the octet header with length - but it would be nice to shove that into a define to make it self-documenting in the three places it is used. I'd also place the 22 later on into a define - nice docs - but macros help to remove the magic number feel.